# 8.4 链接与断点

# 遇到的问题

1. 链接不到自己的ling yi
2. debug断点进不去
3. 最后发现CLion与命令行的行为不一致

# 调试信息

## 调试的断点信息来源

生成的文件中存放有源码路径

msvc->pdb

clang/gcc->dwarf

## 调试断点打不上

1. 调试信息不存在

2. 调试信息中的源码路径对不上

# 通过lldb来尝试

先通过lldb直接启动要调试的内容

[LLDB](https://lldb.llvm.org/use/map.html)

## 查看相关信息

```python
target create "libnncase.dylib"
image list
script(to python)
m = lldb.target.module['image的位置']
for i in range(m.GetNumCompileUnits()):
	cu = m.GetCompileUnitAtIndex(i)
	print(cu)
```

会显示源码路径信息

如果提示日期不同，则打不上断点的问题在于文件未更新

# lldb调试

打断点如果是pending状态，代表在动态加载的时候才会做操作

如果run过一次，后续不会pending，而是直接能打断点

lldb调试断点能够击中则代表生成的库没有问题，个人理解为肯定是实际链接调用的时候调用了以前残留的东西

# 可能残留的位置

全局install的位置

过往编译的库的位置

如果是调用python库，注意看一下有没有pip install过

这里应该用全局搜索，而macOS的这个全局搜索我并不会用，导致这里又多踩了一脚坑

# 删除后重新运行

在这后删除了build生成的dylib，发现命令行如预想的一样报错了，但是CLion依然能够正常运行，证明肯定是哪里存在重复的地方

# 查看python包是从哪里来的

在CLion的python debug console中

```python
import nncase
nncase.__file__
'/Users/fusionbolt/.conda/envs/nncase/lib/python3.8/site-packages/nncase/__init__.py'
```

发现原来是之前调试macOS的时候用pip安装过了...

# 用到的工具（未必用上）

## 查看系统调用

Linux strace

macOS dtruss

## 查看依赖

windows Dependencies

非windows otool -L/-l

# 问题的原因以及自己的疏忽

全局搜索用的不好，其实应该依靠全局搜索搜到conda包安装的nncase中

3.8的python编出了3.7的包，这个问题被我忽略了